test/envtest/telemetryServer.go

```diff
@@ -0,0 +1,128 @@
+type TelemetryServer struct {
+	reportChan chan []byte        // Channel to receive telemetry reports.
+	listener   net.Listener       // TLS listener for incoming telemetry connections.
+	started    bool               // Indicates whether the server is running.
+	cancel     context.CancelFunc // Function to cancel the server's context.
+}
+
+// NewTelemetryServer creates and configures a new TelemetryServer instance.
+// It generates a TLS listener using a self-signed certificate.
+func NewTelemetryServer(t *testing.T) *TelemetryServer {
+	t.Helper()
+	t.Log("configuring TLS listener - server for telemetry data")
+	cert := certificate.MustGenerateCert()
+	telemetryServerListener, err := tls.Listen("tcp", "localhost:0", &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		MinVersion:   tls.VersionTLS13,
+		MaxVersion:   tls.VersionTLS13,
+	})
+	require.NoError(t, err)
+
+	reportChan := make(chan []byte)
+
+	return &TelemetryServer{
+		reportChan: reportChan,
+		listener:   telemetryServerListener,
+	}
+}
+
+// Start begins the telemetry server, accepting incoming connections and processing telemetry data.
+// It does not block.
+func (ts *TelemetryServer) Start(ctx context.Context, t *testing.T) {
+	t.Helper()
+	ctx, cancelFunc := context.WithCancel(ctx)
+
+	// handleConnection processes incoming telemetry data from a single connection.
+	handleConnection := func(ctx context.Context, t *testing.T, conn net.Conn, wg *sync.WaitGroup) {
+		t.Helper()
+		defer func() {
+			if err := conn.Close(); err != nil {
+				t.Logf("error closing connection: %v", err)
+			}
+			wg.Done()
+		}()
+
+		for {
+			report := make([]byte, 2048) // Buffer for telemetry data.
+			n, err := conn.Read(report)
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			if !assert.NoError(t, err) {
+				return
+			}
+			t.Logf("received %d bytes of telemetry report", n)
+			select {
+			case ts.reportChan <- report[:n]: // Send the report to the channel.
+			case <-ctx.Done(): // Exit if the context is canceled.
+				return
+			}
+		}
+	}
+
+	t.Logf("Starting telemetry server")
+	go func() {
+		// Main loop to accept and handle incoming connections.
+		var wg sync.WaitGroup
+		for {
+			select {
+			case <-ctx.Done():
+				t.Logf("Context cancelled, stopping telemetry server")
+				wg.Wait()
+				close(ts.reportChan)
+				t.Logf("Telemetry server stopped")
+				return
+			default:
+				conn, err := ts.listener.Accept()
+				if err != nil && errors.Is(err, net.ErrClosed) {
+					break
+				}
+				if !assert.NoError(t, err) {
+					break
+				}
+				wg.Add(1)
+				go handleConnection(ctx, t, conn, &wg)
+			}
+		}
+	}()
+
+	ts.started = true
+	ts.cancel = cancelFunc
+}
+
+// Endpoint returns the address of the telemetry server.
+func (ts *TelemetryServer) Endpoint() string {
+	return ts.listener.Addr().String()
+}
+
+// ReportChan provides access to the telemetry report channel.
+func (ts *TelemetryServer) ReportChan() chan []byte {
```

Comment:
Can we return a receive only channel here?

It's not meant to send data to right?

```suggestion
func (ts *TelemetryServer) ReportChan() <-chan []byte {
```